rfc-108-including-gem-component-assets.md

```diff
@@ -0,0 +1,66 @@
+# Include specific component assets in applications
+
+## Problem and background
+
+The [current method for including the components](https://github.com/alphagov/govuk_publishing_components/blob/master/docs/install-and-use.md#4-include-the-assets) in a frontend application (such as `government-frontend`) is to include `_all_components.scss` and `_all_components_print.scss` in the Sass of that application. This has some benefits:
+
+- all components are automatically included in the application
+
+This also has some detriments:
+
+- all components are included regardless of whether they are used, so unused CSS may be included
+- only publically facing components can be added to the gem in order to prevent the previous problem
+
+We now have a number of non-public applications that use the components gem and components are being written for those applications. While the past model of putting components into applications has worked, we now have components that need to be shared across more than one private application, which is not possible with the current model.
+
+## Proposed solution
+
+Instead of including all of the component Sass each application should be configured to only include the components that it needs. For example, in an applications `_application.scss`:
+
+```
+// currently
+@import 'govuk_publishing_components/all_components';
+```
+
+```
+// proposed
+@import 'govuk_publishing_components/components/accordion';
+@import 'govuk_publishing_components/components/button';
+// (etc.)
+```
+
+This would mean we can add any components we like to the gem without worrying that they are needlessly adding page weight to the publically facing GOV.UK.
+
+## Javascript
+
+We can do something similar with Javascript.
+
+```
+// currently
+//= require govuk_publishing_components/all_components
+```
+
+```
+// proposed
+//= require govuk_publishing_components/components/accordion
+//= require govuk_publishing_components/components/button
+// (etc.)
```

Comment:
I _think_ the components might need initialising as well?

Reply:
In `all_components.js`, there's a function called that initialises all of the components (`window.GOVUKFrontend.initAll()`). I think that there's an initialisation function for each component that needs to be used to make sure that the component's JS is started.

Reply:
For reference, that function is here:

https://github.com/alphagov/govuk-frontend/blob/5a514ce69db68ec17f681cdec8d82769c9c241b9/src/govuk/all.js#L12-L62

Reply:
Yeah, the JS is more complicated than I've made out here. As a first pass, we could still include all of the JS. It shouldn't cause any problems and won't be worse (in terms of filesizes) than what we already have, and in the meantime work on a better solution.